### ISSUE TITLE
Assertion fails when truncating table with views being built

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#17543]
```
scylla: replica/table.cc:2330: future<db::replay_position> replica::table::discard_sstables(db_clock::time_point): Assertion `std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) { return _compaction_manager.compaction_disabled(cg.as_table_state()); })' failed.
```

```
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
__assert_fail_base.cold at ??:?
__assert_fail at ??:?
replica::table::discard_sstables(std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at table.cc:?
replica::database::truncate(db::system_keyspace&, replica::table&, replica::database::table_truncate_state const&, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::$_0::operator()(view_ptr) const at database.cc:?
replica::database::truncate(db::system_keyspace&, replica::table&, replica::database::table_truncate_state const&, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) [clone .resume] at database.cc:?
seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at main.cc:?
seastar::reactor::run_some_tasks() at ??:?
seastar::reactor::do_run() at ??:?
seastar::reactor::run() at ??:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ??:?
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at ??:?
scylla_main(int, char**) at main.cc:?
std::function<int (int, char**)>::operator()(int, char**) const at ??:?
main at main.cc:?
__libc_start_call_main at ??:?
__libc_start_main_alias_2 at :?
_start at ??:?
```

Found with extension to dtest (PR will come soon)

**Comment by nyh** [comment]
@mykaul FYI I'm swamped in P1 and and P2 issues related to MV, tablets, or both, and don't know which one to look at first (also considering context-switches take a long time) and not getting to other things like generally doing the MV+tablets work or the "MV redesign presentation" you asked about.

**Comment by mykaul** [comment]
> @mykaul FYI I'm swamped in P1 and and P2 issues related to MV, tablets, or both, and don't know which one to look at first (also considering context-switches take a long time) and not getting to other things like generally doing the MV+tablets work or the "MV redesign presentation" you asked about.

We should have discussed it this morning. We can discuss it over email (on specific issues) if we should prioritize / de-prioritize items.

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/pull/17548 by @xemul may or may not fix this issue, I'm waiting for a clarification from him if it fixes it or doesn't, not working on this issue until then.

**Comment by xemul** [comment]
#17548 doesn't fix this issue, unfortunately. #17548 is to fix builder lockup vs tablets migration. I thought that builder would lockup the same vs truncate, but it crashed earlier instead (this issue)

**Comment by nyh** [comment]
> #17548 doesn't fix this issue, unfortunately. #17548 is to fix builder lockup vs tablets migration. I thought that builder would lockup the same vs truncate, but it crashed earlier instead (this issue)

I see. Than given that @mykaul prioritized it P1 and it's in the area of views, I'll work on it.

**Comment by nyh** [comment]
Hi, @raphaelsc you recently modified the assertion in table.cc which failed:
```c++
future<db::replay_position> table::discard_sstables(db_clock::time_point truncated_at) {
    assert(std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) {
        return _compaction_manager.compaction_disabled(cg.as_table_state());
    }));
```
So maybe you remember what it is about?

Couldn't there be a race where some new sstables are created in parallel with truncate, and those might be registered in the compaction manager?

**Comment by raphaelsc** [comment]
> Hi, @raphaelsc you recently modified the assertion in table.cc which failed:
> 
> ```c++
> future<db::replay_position> table::discard_sstables(db_clock::time_point truncated_at) {
>     assert(std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) {
>         return _compaction_manager.compaction_disabled(cg.as_table_state());
>     }));
> ```
> 
> So maybe you remember what it is about?
> 
> Couldn't there be a race where some new sstables are created in parallel with truncate, and those might be registered in the compaction manager?

It can be a problem in the outer layer that doesn't disable compactions properly.

**Comment by nyh** [comment]
A wild guess: Maybe truncate happens on both base and view tables, but compactions are only disabled on the base table and not the view, so we get this failure while truncating the view?
I'll need to add more printouts to make sure. Thanks to @xemul we have a reliable reproducing dtest, `materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_truncate`.

**Comment by raphaelsc** [comment]
compaction_disabled() is implemented like this:

```
bool compaction_manager::compaction_disabled(table_state& t) const {
    return _compaction_state.contains(&t) && _compaction_state.at(&t).compaction_disabled();
}
```

maybe a race of truncate and table drop?

first condition is false if table was deregisted, which only happens on shutdown after table is stopped or table was removed. 

database::truncate_table_on_all_shards() works hard to stop compaction on all tables being truncated, and keeps it like that throughout the truncation. It also does that for any views it might have.

**Comment by raphaelsc** [comment]
We need to exorcise that assert and dump a nicer message instead. If we knew which table failed at that assertion, it could have made the investigation simpler. If it was a view or base.

**Comment by raphaelsc** [comment]
@xemul @nyh  do you have the logs?

**Comment by nyh** [comment]
You're right, I see now that `database::truncate_table_on_all_shards()` indeed looks like it disable compactions for all the views, not just the base table. 
I'll need to add more printouts to see what's wrong. Thanks to @xemul I can reproduce this so I don't need to rely on logs. I'll do this now.

The relevant log ends with:
```
INFO  2024-03-04 20:44:37,966 [shard 0:strm] view - Building view ks.tm_table_mv
_0, starting at token minimum token
INFO  2024-03-04 20:44:37,966 [shard 1:strm] query_processor - Column definition
s for ks.tm_table changed, invalidating related prepared statements
INFO  2024-03-04 20:44:37,972 [shard 0:strm] migration_manager - Gossiping my sc
hema version 40d1aabc-da57-11ee-2063-faebcfc5a4ff
INFO  2024-03-04 20:44:37,973 [shard 0:strm] schema_tables - Schema version chan
ged to 40d1aabc-da57-11ee-2063-faebcfc5a4ff
scylla: replica/table.cc:2330: future<db::replay_position> replica::table::disca
rd_sstables(db_clock::time_point): Assertion `std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) { return _compaction_manager.compaction_disabled(cg.as_table_state()); })' failed.
Aborting on shard 1.
```
The backtrace is obvious and doesn't add much:
```
replica::table::discard_sstables(std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at table.cc:?
replica::database::truncate(db::system_keyspace&, replica::table&, replica::database::table_truncate_state const&, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::$_0::operator()(view_ptr) const at database.cc:?
replica::database::truncate(db::system_keyspace&, replica::table&, replica::database::table_truncate_state const&, std::chrono::time_point<db_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) [clone .resume] at database.cc:?
seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at main.cc:?
```

**Comment by nyh** [comment]
> We need to exorcise that assert and dump a nicer message instead. If we knew which table failed at that assertion, it could have made the investigation simpler. If it was a view or base.

The table which fails the check is the view - tm_table_mv_0. The base table was checked successfully a millisecond earlier.

**Comment by nyh** [comment]
> compaction_disabled() is implemented like this:
> 
> ```
> bool compaction_manager::compaction_disabled(table_state& t) const {
>     return _compaction_state.contains(&t) && _compaction_state.at(&t).compaction_disabled();
> }
> ```
> 
> maybe a race of truncate and table drop?

I don't think this test drops the table (or the view), and I don't see it happening in the log.

> first condition is false if table was deregistered, which only happens on shutdown after table is stopped or table was removed.

I added a printout and the first condition is true - _compaction_state does contain t.
Now I need to figure out how we disabled the compaction on the view (I saw the code), but it wasn't really disabled.

**Comment by raphaelsc** [comment]
> I think what is happening is that a truncate operation starts **before** a view was added, and disables compaction only on the base, and then in parallel a view is created, and when the truncate proceeds it checks and see suddenly we have a view with compaction not disabled!
> 
> Here are a few printouts from the log that show this timeline:
> 
> ```
> INFO  2024-03-04 22:08:11,116 [shard 0:stmt] database - Truncating ks.tm_table with auto-snapshot
> WARN  2024-03-04 22:08:11,116 [shard 0:stmt] database - NYH truncate_table_on_all_shards stopping for base state tm_table
> INFO  2024-03-04 22:08:11,130 [shard 0:strm] schema_tables - Altering ks.tm_table id=e442b050-da62-11ee-a47d-cd381618acdd version=e442b051-da62-11ee-a47d-cd381618acdd
> INFO  2024-03-04 22:08:11,130 [shard 0:strm] schema_tables - Creating ks.tm_table_mv_0 id=ece561d0-da62-11ee-b35e-a132135dd95a version=ece561d1-da62-11ee-b35e-a132135dd95a
> INFO  2024-03-04 22:08:11,131 [shard 0:strm] view - Building view ks.tm_table_mv_0, starting at token minimum token
> WARN  2024-03-04 22:08:11,151 [shard 1:stmt] table - NYH discard_sstables tm_table_mv_0
> ```
> 
> As you can see, in millisecond 116 the truncation started and compaction was disabled only on the base, in millisecond 130 the view was added, and only in millisecond 151 the discard went to check if the views all have disabled compactions, and they didn't.
> 
> I don't know what the solution should be. Note that between millisecond 131 and 151, view building ran and might have filled data in the view, so it is necessary to truncate the view too - it's not enough to skip it, we really need to truncate it :-(
> 
> Why was it important to disable compaction of both base and all views as a first step and only then truncate them all? Can't we disable compaction for the base, then truncate the base, and only then disable compaction for the view and then truncate the view?

I think so, but let's also listen to @bhalevy as he's well versed in this area. Note that if we do that, we can skip views added later (to not bump into the assert), as by the time we truncate views, we know base table is truncated already.

**Comment by raphaelsc** [comment]
but base and views are not necessarily co-located on the same shard or node. do we possibly need to coordinate it globally? 

first we make sure all base tables are truncated first on all nodes (using something like a barrier), before we proceed to truncating the views.

otherwise base table not truncated yet in node 1 could still push data to view added later in node 2, right?

**Comment by raphaelsc** [comment]
> but base and views are not necessarily co-located on the same shard or node. do we possibly need to coordinate it globally?
> 
> first we make sure all base tables are truncated first on all nodes (using something like a barrier), before we proceed to truncating the views.
> 
> otherwise base table not truncated yet in node 1 could still push data to view added later in node 2, right?

an example of how it might happen:

t0: truncate is issued on base table
t1: node 2 truncates base table and views
t2: new view is created
t3: node 1 hasn't truncated base yet
t4: new view starts being built with data from node 1
t5: node 1 finishes truncation of base and views
t6: new view in node 2 wasn't truncated, and has stale data

**Comment by nyh** [comment]
> but base and views are not necessarily co-located on the same shard or node. do we possibly need to coordinate it globally?

The existence or non-existence of some view is in the schema which is (sort-of) global.

> first we make sure all base tables are truncated first on all nodes (using something like a barrier), before we proceed to truncating the views.
> 
> otherwise base table not truncated yet in node 1 could still push data to view added later in node 2, right?

Maybe what we need to do is to stop view building for this base-table as the first step? This stopping-of-view-building will also apply to new views (if they appear). 

But this alone will not solve the problem we have now, with the assertion failure... The assertion failure happens because the test creats a new view between the time that the truncate started and didn't mark this view (that didn't yet exist) for no-compaction, and the time that the truncate decided to look again at the list of views and found the new view - and was shocked to see it didn't have compaction disabled.

I'm afraid I don't fully understand what was the intention for doing this compaction-disable thing in advance. When we see a view that doesn't have compaction disabled, why can't we disable it right there? Why did we need to disable it in advance? What risk is this trying to avoid?

**Comment by nyh** [comment]
> t0: truncate is issued on base table 
> t1: node 2 truncates base table and views 
> t2: new view is created
> t3: node 1 hasn't truncated base yet
> t4: new view starts being built with data from node 1
> t5: node 1 finishes truncation of base and views
> t6: new view in node 2 wasn't truncated, and has stale data

How does the existing code attempt to avoid this situation? How does the "disable compaction" thing help that?

It sounds to me like the right thing to do something like the following:
1. If the base table is currently building any views, stop the view building (on all nodes) and mark them completed (I'm not sure that'll be easy, but it should be doable).
2. Truncate the base table and wait for this truncation to be complete on all nodes.
3. Truncate the view table and wait for this truncation.

One might argue that this is not atomic: It is possible that between step 2 and 3, someone might send data and it will be written to the base, and to the view, but then be truncated from the view. I admit that's true. But the current implementation - how is it atomic? Was the approach we used in truncate for views ever documented anywhere?

**Comment by raphaelsc** [comment]
> > t0: truncate is issued on base table
> > t1: node 2 truncates base table and views
> > t2: new view is created
> > t3: node 1 hasn't truncated base yet
> > t4: new view starts being built with data from node 1
> > t5: node 1 finishes truncation of base and views
> > t6: new view in node 2 wasn't truncated, and has stale data
> 
> How does the existing code attempt to avoid this situation? How does the "disable compaction" thing help that?
> 

I think it's only there to prevent compaction from mixing data-to-be-truncated with data-to-not-be-truncated (written while truncate is happening).

**Comment by bhalevy** [comment]
@nyh If and when we fix https://github.com/scylladb/scylladb/issues/11230 the truncate tombstone should apply the same authoritative write timestamp to both base tables and its views.
With that we shouldn't need the hacks against resurrecting data since the truncate tombstone should be garbage collected only when there's no data with older write timestamp existing in the sstables, commitlog, or memtable.

Cc @michoecho

**Comment by nyh** [comment]
I think this issue is clear, and less serious or urgent than I thought earlier, so I am removing my self-assignment.

For the future me or someone else who will work on this issue, here is a summary of what we know so far, so there's no need to read the above 23 comments:

1. The issue can be easily reproduced, by the dtest `materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_truncate` that fails every time.
2. The bug happens when a CREATE MATERIALIZED VIEW is created in parallel to a TRUNCATE operation.
3. The bug timeline is as follows: TRUNCATE calls `database::truncate_table_on_all_shards() ` and begins by disabling compaction on the table and all its views. If a few milliseconds later a new view is created, it will not have compaction disabled. Later, TRUNCATE again lists the table's views to run `table::discard_sstables()` on each of them - and the new list will contain the new view. When `table::discard_sstables()` is called on the new view, it asserts that compaction is disabled for this table (a view) - which it isn't because it didn't yet exist when we were disabling compaction. 

There's a bigger question of how atomic a truncation of a table with views is today, and how it should be, but I don't think it should be solved for this issue (perhaps its solution should be https://github.com/scylladb/scylladb/issues/11230). 

For this issue we just need the assertion failure to go away. **Perhaps even just removing the assertion will be a good-enough solution**: disabling compactions helps avoid one kind of data resurrection problem, but truncating a view has other problems (e.g., what if view building continues to fill the truncated view with data?) so I don't think that insisting on disabling compactions on a view that was just created is the most important of these problems.

**Comment by nyh** [comment]
@xemul we have another problem :-( The dtest that you wrote (`materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_truncate`) will start running as soon as this issue is marked fixed (due to a "require"). But, if I fix this issue by simply removing the assert(), the test actually fails:

```
View ks.tm_table_mv_1 not built
```

I think this might be fixed by https://github.com/scylladb/scylladb/pull/17548 (in fact I think you came up with this test case that I'm fixing here while working on that other pull request). So we'll need to merge that fix before this one.

**Comment by xemul** [comment]
> I think this might be fixed by https://github.com/scylladb/scylladb/pull/17548

Well, it indeed helps fixing the `View ks.tm_table_mv_1 not built` just to step on another one:

```
FAILED materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_truncate - AssertionError: Expected a row count of 40000 in table 'tm_table', but got 0
```

**Comment by xemul** [comment]
Fixing this is simple, table is truncated, so exected row  count should be 0. Like this

```patch
--- a/materialized_views_test.py
+++ b/materialized_views_test.py
@@ -917,6 +916,7 @@ class TestMaterializedViews(CommonUtils):
             rows_after_test = max(0, prefill - (max_delete - 1000))
         elif change_type == "truncate":
             change_func = {"func": tm.truncate_table}
+            rows_after_test = 0
         elif change_type == "add node":
             change_func = {"func": self._add_new_node, "kwargs": {"delay": 1}}
         elif change_type == "decommission":
```

so the next failure is

```
AssertionError: Expected a row count of 0 in table 'tm_table_mv_0', but got 6784
```

this looks real to me -- table is truncated, but view is not

**Comment by nyh** [comment]
> this looks real to me -- table is truncated, but view is not

Unfortunately, this will be a real problem which my patch doesn't fix (because it has nothing to do with **compaction**). The truncation is not atomic between a table and its views: We could be truncating the view first, but the base table still exists and view building is running and writing data to the view - and then we truncate the base.

I think this should be tolerated by this test - and a new test, and a new issue - should be created to report this problem. We will need to merge something at some point, and don't need to solve all the problems in this area in one patch.
 
By the way, @fruch this is why I was always against the idea of "requires" and preferred the manual "xfail" approach. There's a high probability that a known-failing test will **not** actually start passing when a fix is committed, because the test was always buggy and we didn't notice. In the cql-pytest world (where we use "xfail", not "requires"), re-enabling the test would require a patch - sometimes just removing the "xfail" but unfortunately not rarely it includes also a fix to the test together with the removal of the xfail.

**Comment by xemul** [comment]
> Unfortunately, this will be a real problem which my patch doesn't fix (because it has nothing to do with compaction)

Absolutely

> I think this should be tolerated by this test - and a new test, and a new issue - should be created to report this problem. We will need to merge something at some point, and don't need to solve all the problems in this area in one patch.

Yes, definitely. I already submitted the "fix N 2" -- make dtest expect no rows from base table scylladb/scylla-dtest#4029. As far as the expected rows for view table is concerned -- do you want me to create the issue?

And also, what about xfail/requires for the test itself? Initially I marked it as skipped, I don't like requires decoration too, but Israel aksed for requires one.

**Comment by nyh** [comment]
> Yes, definitely. I already submitted the "fix N 2" -- make dtest expect no rows from base table [scylladb/scylla-dtest#4029](https://github.com/scylladb/scylla-dtest/pull/4029). As far as the expected rows for view table is concerned -- do you want me to create the issue?

I'll do this now
 
> And also, what about xfail/requires for the test itself? Initially I marked it as skipped, I don't like requires decoration too, but Israel aksed for requires one.

I'll create the new issue and then ask you in your above patch for dtest to add another "requires" (is it possible to have two? if not, just replace it) for the new issue that won't be solved today.

**Comment by fruch** [comment]
> 
> > Yes, definitely. I already submitted the "fix N 2" -- make dtest expect no rows from base table [scylladb/scylla-dtest#4029](https://github.com/scylladb/scylla-dtest/pull/4029). As far as the expected rows for view table is concerned -- do you want me to create the issue?
> 
> I'll do this now
>  
> > And also, what about xfail/requires for the test itself? Initially I marked it as skipped, I don't like requires decoration too, but Israel aksed for requires one.
> 
> I'll create the new issue and then ask you in your above patch for dtest to add another "requires" (is it possible to have two? if not, just replace it) for the new issue that won't be solved today.
> 

Yes it possible to have two require marks
I've fixed that few months back
